### ISSUE TITLE
Using mypy via pre-commit

### ISSUE THREAD

**Issue by hauntsaninja** [python/mypy#13916]
pre-commit is useful; many mypy users use mypy via pre-commit. This issue exists so I can link to it when users run into common gotchas. Please do not discuss pre-commit related feature requests here, open a new issue.

### pre-commit runs mypy in an isolated environment

pre-commit wants checks to run in isolated environments. mypy wants to know as much about your program as it can, including everything about everything you import. These are in conflict.

The pre-commit mypy mirror thing automatically (and to most users, surprisingly) passes `--ignore-missing-imports`, which means that most types originating from third party dependencies will become `Any`: https://github.com/pre-commit/mirrors-mypy/blob/33f4a30be4e66513c51a857712fc732e1a9ddd78/.pre-commit-hooks.yaml#L7
In my opinion, this is **really bad**: it will silently make type checking far less effective. It also means you'll likely get different results when running mypy directly.

Two possible pre-commit fixes in your pre-commit config:
1) Use `language: system` (and maybe a helper script) to run something from outside a pre-commit environment
2) Repeat your list of dependencies (and stub dependencies) in `additional_dependencies`

### pre-commit passes only diffed files

This is sort of the core idea of pre-commit and for most pre-commit checks, this is great. But mypy really wants to do whole program analysis, and will try to follow all imports, so pre-commit is not actually saving you any work. In fact, if it is saving you work, it's probably undesirable and some bug in your mypy setup. Anyway, this is another source of potential differences between running mypy directly, e.g. there are undesirable interactions with `--exclude`: mypy assumes that if you pass a file in the command, you want it to be checked. 

Recommended pre-commit fix:
In your pre-commit config, use `pass_filenames: false`

**Comment by alex-shef** [comment]
There is the error in such case: "Executable `mypy` not found"

**Comment by Jamie-BitFlight** [comment]
For those landing here looking for a solution to the `additional_dependencies` problem, especially when linting scripts with PEP 723 inline dependencies:

I built `pep723-loader` which wraps mypy (or any linter) and auto-installs PEP 723 dependencies before execution:

```yaml
- repo: local
  hooks:
    - id: mypy
      name: mypy
      entry: uv run -q --no-sync --with pep723-loader --with mypy pep723-loader mypy
      language: system
      types: [python]
      pass_filenames: true
```

It scans passed files for `# /// script` metadata blocks, extracts dependencies via `uv export --script`, installs them, then runs mypy. Exit codes propagate correctly.

This addresses the "language: system" workaround mentioned above while handling PEP 723 scripts that declare their own inline dependencies.

PyPI: `pip install pep723-loader`
Repo: https://github.com/bitflight-devops/pre-commit-pep723-linter-wrapper

**Comment by ilevkivskyi** [comment]
It looks like people continue hitting issues with pre-commit. Maybe we should keep this issue open (and maybe even make it pinned)? Another thing, maybe we can convince whoever maintains `mirrors-mypy` to include `pass_filenames: false` as a default etc?

**Comment by ilevkivskyi** [comment]
Btw another thing (from one of the issues caused by this):
> The responsible alternative would be to write a pid in a lockfile in the cache and raise an exception as soon as a parallel run is detected.

I was thinking about this for a while, maybe we should actually do this? It will probably break builds for many people, but at least it will fail deterministically, and we can give a clear error message explaining what to do.

These problems are going to happen much more often with SQLite cache cc @JukkaL

**Comment by ilevkivskyi** [comment]
Btw we can probably close https://github.com/python/mypy/issues/14521 in favor of this issue, and update the description to mention the random crashes from parallel runs (and a solution like `require_serial: true`). Most crashes from parallel runs are likely from `pre-commit` anyway.

**Comment by dvarrazzo** [comment]
Some experience gathered with psycopg codebase:

- using `language: system` is useful to guarantee consistency between the dev environment and the ci environment and to specify dependencies in a single place.
- normal behaviour results in parallel runs which are considered an issue, and become a practical problem when using the SQLite cache (on by default in Mypy 1.20, see #21136).
- disabling parallel runs by using `require_serial: false` results in all the filenames passed to mypy when running `pre-commit -a`, which may cause issues such as #17362.
- running mypy on the entire codebase instead of just the files changed, by using `pass_filenames: false` seems slower for interactive usage (on psycopg on my dev box it results in a ~25% penalty when changing a single file, I haven't tested further).

I assume that `pass_filenames: false` is the safest configuration. For the moment I will personally try to maintain the 1.19 behaviour by passing `--no-sqlite-cache`, but I'll keep an eye here to see if the conventional wisdom shifts.

**Comment by JukkaL** [comment]
What about serializing mypy runs via SQLite cache? If we detect that another mypy pid is using the cache, we could just wait until the cache is available again. This could slow down performance but it would be correct.

Another idea would be to detect that the SQLite cache is being used by another process, and then automatically try using a different name for the cache (e.g. `cache-2.db`, `cache-3.db`, etc.) that isn't locked. I think this would both fix the issue and provide reasonable performance, at least if the cache is persisted across runs. However, the cache would use more disk space.

**Comment by ilevkivskyi** [comment]
@JukkaL 
> What about serializing mypy runs via SQLite cache? If we detect that another mypy pid is using the cache, we could just wait until the cache is available again. This could slow down performance but it would be correct.

I don't think this can be correct unless you lock for ~entire duration of the run, especially if different invocations have different options and/or source sets. The problem is that we validate cache metas during the initial graph loading, while cache data files are loaded during stale SCC processing. And no, re-validate metas just before processing isn't really an option, because this will complicate many things (while `build.py` is already complex), especially in case of parallel checking, where we also rely on cache metas to "broadcast" interface hashes.
> Another idea would be to detect that the SQLite cache is being used by another process, and then automatically try using a different name for the cache (e.g. `cache-2.db`, `cache-3.db`, etc.) that isn't locked.

FWIW I don't like this either, for two reasons:
* There is one valid use case for running multiple mypy invocations in parallel - when you have several completely (or mostly) disjoint projects in the same repo. However, in this case having multiple cache DBs will make incremental mode completely ineffective, as the next mypy run will likely to pick up a "wrong" cache. As I mentioned in other issues, our recommendation should be to use _stable_ `--cache-dir` for each invocation.
* This solution is SQLite-specific. Although this problem is mostly affects SQLite cache, it can happen with FS cache as well, just less likely.

**Comment by ilevkivskyi** [comment]
Another possible option is to add a flag (say `--force-cache-isolation`) that would give a detailed error if multiple mypy invocations accessing same `--cache-dir` are detected. It would be on by default, and if it is disabled, it would fall back to per-invocation entire-duration locking. This sound like an overkill however.

It looks like there are no good options so far, so we may be forced to choose the least bad one.

**Comment by JukkaL** [comment]
Here are a few more options to consider. @msullivan came up with these ideas:

* If another mypy run is using the cache, create a temporary 'overlay' cache db which will fall back to the primary cache for reads that can't be found in the overlay cache. So all writes would go to the overlay cache. Delete the overlay cache at the end of a run.
* Automatically turn off caching for any parallel runs.

Here's one more idea, which is a simplified variation of the overlay idea:
* If another mypy run is using the cache, any parallel runs will use read only access to the cache, i.e. they can use existing cached data, but can't write any new data. This should still help with cached stdlib modules etc.

I kind of like the last idea of giving parallel runs read-only access, if it's something we can implement easily, as it sounds easy to reason about and doesn't seem to have really bad tradeoffs such as cache file size explosion.

If we mplement such locking, we'll need to make sure the lock doesn't persist indefinitely if a mypy process gets killed using SIGKILL (no cleanup). Apparently sqlite uses OS file locks so they should get released correctly on SIGKILL if we can use the built-in sqlite exclusion mechanism.

**Comment by ilevkivskyi** [comment]
I think this is also wrong. We are not really trying to solve some puzzle here. The real problem is that if a user runs multiple mypy invocations in parallel, then most likely they are doing something wrong. None of these solutions are actually helping the user. After all, what I think we should do is:
1. Add some docs about recommended way to use mypy with pre-commit. Essentially, just a more formal version of the OP plus `require_serial: true`.
2. Add a short docs subsection (e.g. to common issues section) about supported use case for running multiple mypy's in parallel, and how to do this safely, i.e. always use a stable `--cache-dir` for each "disjoint" project (this subsection would also link to the one above).
3. Detect the parallel use of the same cache, and give a _non-blocking_ error linking to these two docs sections. Then continue with some best-effort fallback behaviour (like disabling the cache).

If we do it this way, then the exact fallback-behavior is less important.

**Comment by srittau** [comment]
As a data point: We're also running into this in typeshed: https://github.com/python/typeshed/actions/runs/25555397598/job/75013016030?pr=15716

We're already using `--no-incremental` (with a comment referencing #11220). Since I don't think that running multiple instances of mypy at the same time is that unreasonable, I would suggest printing an error message pointing to `--cache-dir`, at least as a short-term solution.

**Comment by ilevkivskyi** [comment]
@srittau Actually I think you will likely not need `--no-incremental` if you set a separate `--cache-dir` for each invocation.

In general I think we should go ahead with the plan I proposed in https://github.com/python/mypy/issues/13916#issuecomment-4179149990

**Comment by ilevkivskyi** [comment]
The problems caused by this keep coming. FWIW I am going to mark this as high priority, and pin the issue for now.
